### PR TITLE
Add command to unregister adherents

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/Command/UnregistrerAdherentsCommand.php [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+     * @var AdherentRepository
+     */
+    private $adherentRepository;
+
+    /**
+     * @var AdherentManager
+     */
+    private $adherentManager;
+
+    private $emailNotFound;
```
Comment (HeahDude) [inline src/Command/UnregistrerAdherentsCommand.php]: `$notFoundEmails`

File: src/Command/UnregistrerAdherentsCommand.php [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        $this->adherentManager = $this->getContainer()->get('app.adherent_manager');
+        $this->emailNotFound = [];
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        try {
+            $emails = $this->parseCSV($input->getArgument('fileUrl'), 'r');
+        } catch (FileNotFoundException $exception) {
+            $output->writeln(
```
Comment (HeahDude) [inline src/Command/UnregistrerAdherentsCommand.php]: `$output->writeln($exception->getMessage());`

File: src/Command/UnregistrerAdherentsCommand.php [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+        if ($this->unregistrateAdherents($emails, $output)) {
+            return 1;
+        }
+
+        $this->em->commit();
+
+        $output->writeln('Unregistration finished.');
+
+        if (0 < count($this->emailNotFound)) {
```
Comment (HeahDude) [inline src/Command/UnregistrerAdherentsCommand.php]: `if ($this->emailNotFound) {`, same below.

File: src/Command/UnregistrerAdherentsCommand.php [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+            $output->writeln('The following email adress are not found in DB :');
+            foreach ($this->emailNotFound as $email) {
+                $output->writeln($email);
+            }
+        }
+        if (0 < count($this->hosts)) {
+            $output->writeln('The adherents with following email adress are hosts of committees. They are not unregistrated :');
```
Comment (HeahDude) [inline src/Command/UnregistrerAdherentsCommand.php]: unregistrated => unregistered. Same below.

File: src/Command/UnregistrerAdherentsCommand.php [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+use AppBundle\Repository\AdherentRepository;
+use Doctrine\ORM\EntityManager;
+use Symfony\Bundle\FrameworkBundle\Command\ContainerAwareCommand;
+use Symfony\Component\Console\Helper\ProgressBar;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Filesystem\Exception\FileNotFoundException;
+
+class UnregistrerAdherentsCommand extends ContainerAwareCommand
```
Comment (HeahDude) [inline src/Command/UnregistrerAdherentsCommand.php]: Typo: Unregistrer => Unregister

File: src/Command/UnregistrerAdherentsCommand.php [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    private $hosts;
+
+    private $citizenProjectAdministrators;
+
+    protected function configure()
+    {
+        $this
+            ->setName(self::COMMAND_NAME)
+            ->addArgument('fileUrl', InputArgument::REQUIRED)
+            ->setDescription('Unregistrate adherents from CSV file (only the fifth column "email" is taken into account)');
```
Comment (HeahDude) [inline src/Command/UnregistrerAdherentsCommand.php]: Unregistrate => Unregister

File: src/Command/UnregistrerAdherentsCommand.php [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            $output->writeln(
+                sprintf('%s file not found', $input->getArgument('fileUrl'))
+            );
+
+            return 1;
+        }
+
+        $this->em->beginTransaction();
+
+        if ($this->unregistrateAdherents($emails, $output)) {
```
Comment (HeahDude) [inline src/Command/UnregistrerAdherentsCommand.php]: `unregistrateAdherents` => `unregisterAdherents`

File: src/Membership/AdherentManager.php [hunk 8]
```diff
@@ -2,25 +2,63 @@
...
+        CommitteeManager $committeeManager,
+        CommitteeFeedManager $committeeFeedManager,
+        ActivitySubscriptionManager $activitySubscriptionManager,
+        CitizenProjectManager $citizenProjectManager,
+        ReportManager $reportManager,
+        EventRegistrationManager $registrationManager
     ) {
         $this->repository = $repository;
         $this->manager = $manager;
+        $this->citizenInitiativeManager = $citizenInitiativeManager;
```
Comment (HeahDude) [inline src/Membership/AdherentManager.php]: All those dept for one method, I would create another service instead, something like `AdherentRegistry::unregister()`

File: src/Command/UnregistrerAdherentsCommand.php [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    private function unregistrateAdherents(array $emails, OutputInterface $output): bool
+    {
+        $batchSize = 20;
+        $i = 1;
+        $progress = new ProgressBar($output, count($emails));
+        $progress->start();
+
+        foreach ($emails as $email) {
+            try {
+                if (null === ($adherent = $this->adherentRepository->findByEmail($email))) {
```
Comment (HeahDude) [inline src/Command/UnregistrerAdherentsCommand.php]: Not related, but this method is badly named, should be called `findOneByEmail` to respect Doctrine conventions and avoid confusion.

File: src/Command/UnregistrerAdherentsCommand.php [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+                if ($adherent->isAdministrator()) {
+                    $this->citizenProjectAdministrators[] = $email;
+
+                    continue;
+                }
+
+                $this->adherentManager->unregister($adherent, false);
+            } catch (\Exception $ex) {
+                $output->writeln(
+                    sprintf('An error occured while unregistering an adherent with email %s : %s', $email, $ex->getMessage())
```
Comment (HeahDude) [inline src/Command/UnregistrerAdherentsCommand.php]: You could put some quotes around the email, no space before the colon.

File: src/Command/UnregistrerAdherentsCommand.php [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+            ++$i;
+        }
+
+        $this->em->flush(); //Persist objects that did not make up an entire batch
+        $this->em->clear();
+
+        $progress->finish();
+
+        return false;
```
Comment (HeahDude) [inline src/Command/UnregistrerAdherentsCommand.php]: Returning false to say that everything went ok is a bit counter intuitive. This private method should not return a boolean but throw a runtime exception instead of returning true.

File: src/Command/UnregistrerAdherentsCommand.php (hunk 5 above)
Comment (illusionOfParadise) [reply src/Command/UnregistrerAdherentsCommand.php]: Done

File: src/Command/UnregistrerAdherentsCommand.php (hunk 1 above)
Comment (illusionOfParadise) [reply src/Command/UnregistrerAdherentsCommand.php]: Done

File: src/Command/UnregistrerAdherentsCommand.php (hunk 6 above)
Comment (illusionOfParadise) [reply src/Command/UnregistrerAdherentsCommand.php]: Done

File: src/Command/UnregistrerAdherentsCommand.php (hunk 2 above)
Comment (illusionOfParadise) [reply src/Command/UnregistrerAdherentsCommand.php]: Done

File: src/Command/UnregistrerAdherentsCommand.php (hunk 7 above)
Comment (illusionOfParadise) [reply src/Command/UnregistrerAdherentsCommand.php]: Done

File: src/Command/UnregistrerAdherentsCommand.php (hunk 3 above)
Comment (illusionOfParadise) [reply src/Command/UnregistrerAdherentsCommand.php]: Done

File: src/Command/UnregistrerAdherentsCommand.php (hunk 4 above)
Comment (illusionOfParadise) [reply src/Command/UnregistrerAdherentsCommand.php]: Done

File: src/Command/UnregistrerAdherentsCommand.php (hunk 9 above)
Comment (illusionOfParadise) [reply src/Command/UnregistrerAdherentsCommand.php]: Done

File: src/Command/UnregistrerAdherentsCommand.php (hunk 10 above)
Comment (illusionOfParadise) [reply src/Command/UnregistrerAdherentsCommand.php]: Done

File: src/Command/UnregistrerAdherentsCommand.php (hunk 11 above)
Comment (illusionOfParadise) [reply src/Command/UnregistrerAdherentsCommand.php]: Done

File: src/Membership/AdherentManager.php (hunk 8 above)
Comment (illusionOfParadise) [reply src/Membership/AdherentManager.php]: Done

File: src/Membership/AdherentRegistry.php [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    private $committeeManager;
+    private $committeeFeedManager;
+    private $activitySubscriptionManager;
+    private $citizenProjectManager;
+    private $reportManager;
+    private $registrationManager;
+
+    public function __construct(
+        EntityManagerInterface $em,
+        ObjectManager $manager,
```
Comment (HeahDude) [inline src/Membership/AdherentRegistry.php]: Why is this needed? You already have an `EntityManagerInterface` that should be enough.

File: app/config/services/platform.xml [hunk 13]
```diff
@@ -60,10 +60,8 @@
         </service>
 
         <!-- Adherents -->
-        <service id="app.adherent_manager" class="AppBundle\Membership\AdherentManager">
-            <argument type="service" id="app.repository.adherent"/>
-            <argument type="service" id="doctrine.orm.entity_manager"/>
-        </service>
+        <service autowire="true" id="app.adherent_manager" class="AppBundle\Membership\AdherentManager" />
+        <service autowire="true" id="AppBundle\Membership\AdherentRegistry" />
```
Comment (HeahDude) [inline app/config/services/platform.xml]: So you can move these two definitions in `services.xml`.

File: src/Command/UnregisterAdherentsCommand.php [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+        $this->em->beginTransaction();
+
+        $this->unregisterAdherents($emails, $output);
+
+        $this->em->commit();
+
+        $output->writeln(PHP_EOL.'Unregistration finished.');
+
+        if ($this->notFoundEmails) {
+            $output->writeln('The following email adresses are not found in DB :');
```
Comment (HeahDude) [inline src/Command/UnregisterAdherentsCommand.php]: You have a typo in all following messages: adress => address

File: src/Repository/AdherentRepository.php [hunk 15]
```diff
@@ -50,7 +50,7 @@ public function count(): int
      *
      * @return Adherent|null
      */
-    public function findByEmail(string $email)
+    public function findOneByEmail(string $email)
```
Comment (HeahDude) [inline src/Repository/AdherentRepository.php]: `: ?Adherent`

File: src/Command/ImportBoardMemberCommand.php [hunk 16]
```diff
@@ -95,9 +95,9 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $this->em->commit();
 
         $output->writeln('Import finish');
-        if (0 < count($this->emailNotFound)) {
+        if ($this->notFoundEmails) {
             $output->writeln('The following email adress are not found in DB');
```
Comment (B-Galati) [inline src/Command/ImportBoardMemberCommand.php]: typo unrelated to your PR :
~adress~|addresses
~are~|were

File: src/Command/UnregisterAdherentsCommand.php [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+                if ($adherent->isAdministrator()) {
+                    $this->citizenProjectAdministrators[] = $email;
+
+                    continue;
+                }
+
+                $unregistration = $unregistrationFactory->createFromAdherent($adherent);
+                $this->adherentRegistry->unregister($adherent, $unregistration);
+            } catch (\Exception $ex) {
+                throw new \RuntimeException(PHP_EOL.sprintf('An error occured while unregistering an adherent with email "%s": %s', $email, $ex->getMessage()));
```
Comment (B-Galati) [inline src/Command/UnregisterAdherentsCommand.php]: Please add `$ex` to the constructor as the `$previous` exception

File: src/Membership/UnregistrationFactory.php [hunk 18]
```diff
@@ -19,4 +19,17 @@ public function createFromUnregistrationCommandAndAdherent(UnregistrationCommand
 
         return $unregistration;
     }
+
+    public function createFromAdherent(Adherent $adherent): Unregistration
```
Comment (HeahDude) [inline src/Membership/UnregistrationFactory.php]: This factory is not needed, since it has no deps, those methods could just be static and live in the `Unregistration` class.

File: src/Command/UnregisterAdherentsCommand.php [hunk 19]
```diff
@@ -118,6 +119,7 @@ private function unregisterAdherents(array $emails, OutputInterface $output): vo
     {
         $batchSize = 20;
         $i = 1;
+        $unregistrationFactory = new UnregistrationFactory();
```
Comment (HeahDude) [inline src/Command/UnregisterAdherentsCommand.php]: Could be removed then.

File: src/Command/UnregisterAdherentsCommand.php [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+        $this->em->beginTransaction();
+
+        $this->unregisterAdherents($emails, $output);
+
+        $this->em->commit();
+
+        $output->writeln(PHP_EOL.'Unregistration finished.');
+
+        if ($this->notFoundEmails) {
+            $output->writeln('The following email addresses are not found in DB :');
```
Comment (B-Galati) [inline src/Command/UnregisterAdherentsCommand.php]: ~are~|were

File: src/Membership/UnregistrationFactory.php (hunk 18 above)
Comment (illusionOfParadise) [reply src/Membership/UnregistrationFactory.php]: Done

File: src/Command/UnregisterAdherentsCommand.php (hunk 19 above)
Comment (illusionOfParadise) [reply src/Command/UnregisterAdherentsCommand.php]: Done

File: src/Command/ImportBoardMemberCommand.php (hunk 16 above)
Comment (illusionOfParadise) [reply src/Command/ImportBoardMemberCommand.php]: Done

File: src/Command/UnregisterAdherentsCommand.php (hunk 20 above)
Comment (illusionOfParadise) [reply src/Command/UnregisterAdherentsCommand.php]: Done

File: src/Command/UnregisterAdherentsCommand.php (hunk 17 above)
Comment (illusionOfParadise) [reply src/Command/UnregisterAdherentsCommand.php]: Done

File: src/Command/UnregisterAdherentsCommand.php [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+        $this->em->commit();
+
+        $output->writeln(PHP_EOL.'Unregistration finished.');
+
+        if ($this->notFoundEmails) {
+            $output->writeln('The following email addresses were not found in DB :');
+            foreach ($this->notFoundEmails as $email) {
+                $output->writeln($email);
+            }
+        }
```
Comment (thomasbisignani) [inline src/Command/UnregisterAdherentsCommand.php]: Just a little note, can you add an empty line after a `if` condition ?

File: src/Command/UnregisterAdherentsCommand.php [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+        if ($this->hosts) {
+            $output->writeln('The adherents with following email address are hosts of committees. They are not unregistered :');
+            foreach ($this->hosts as $email) {
+                $output->writeln($email);
+            }
+        }
```
Comment (thomasbisignani) [inline src/Command/UnregisterAdherentsCommand.php]: Same here, can you add an empty line ?

File: src/Membership/AdherentRegistry.php [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+        $this->eventManager->removeOrganizerEvents($adherent);
+        $this->registrationManager->anonymizeAdherentRegistrations($adherent);
+        $this->committeeFeedManager->removeAuthorItems($adherent);
+        $this->activitySubscriptionManager->removeAdherentActivities($adherent);
+        $this->citizenProjectManager->removeAuthorItems($adherent);
+        $this->reportManager->anonymAuthorReports($adherent);
+
+        if ($token) {
+            $this->em->remove($token);
+        }
```
Comment (thomasbisignani) [inline src/Membership/AdherentRegistry.php]: Same here

Comment (illusionOfParadise) [reply src/Membership/AdherentRegistry.php]: Done

File: src/Command/UnregisterAdherentsCommand.php (hunk 22 above)
Comment (illusionOfParadise) [reply src/Command/UnregisterAdherentsCommand.php]: Done

File: src/Command/UnregisterAdherentsCommand.php (hunk 21 above)
Comment (illusionOfParadise) [reply src/Command/UnregisterAdherentsCommand.php]: Done
